[IsDeprecatedWeakRefSmartPointerException] Make more WebAuthentication classes RefCounted

basuke opened the pull request:
#### 4de859dce063802780052a99ac2045701e391853
<pre>
[IsDeprecatedWeakRefSmartPointerException] Make more WebAuthentication classes RefCounted
<a href="https://bugs.webkit.org/show_bug.cgi?id=281637">https://bugs.webkit.org/show_bug.cgi?id=281637</a>

Reviewed by Chris Dumez.

HidConnection class and its subclasses in WebAuthentication are managed as WeakPtr so they should be ref counted.

* Source/WebKit/UIProcess/WebAuthentication/Cocoa/HidConnection.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/HidConnection.mm:
(WebKit::HidConnection::create):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/HidService.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/HidService.mm:
(WebKit::HidService::createHidConnection const):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidConnection.cpp:
(WebKit::MockHidConnection::create):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidConnection.h:
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidService.cpp:
(WebKit::MockHidService::createHidConnection const):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidService.h:
* Source/WebKit/UIProcess/WebAuthentication/Virtual/VirtualHidConnection.cpp:
(WebKit::VirtualHidConnection::create):
* Source/WebKit/UIProcess/WebAuthentication/Virtual/VirtualHidConnection.h:
* Source/WebKit/UIProcess/WebAuthentication/Virtual/VirtualService.mm:
(WebKit::VirtualService::startDiscoveryInternal):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapHidDriver.cpp:
(WebKit::CtapHidDriverWorker::CtapHidDriverWorker):
(WebKit::CtapHidDriverWorker::~CtapHidDriverWorker):
(WebKit::CtapHidDriverWorker::transact):
(WebKit::CtapHidDriverWorker::write):
(WebKit::CtapHidDriverWorker::reset):
(WebKit::CtapHidDriverWorker::cancel):
(WebKit::CtapHidDriver::CtapHidDriver):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapHidDriver.h:
(WebKit::CtapHidDriverWorker::protectedConnection):

Canonical link: <a href="https://commits.webkit.org/285345@main">https://commits.webkit.org/285345@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ae8dd2721583f56463730ccd8a8b0bde3ca649e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72228 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51649 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25016 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76396 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23437 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59453 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23259 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56932 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15440 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75295 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46790 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62200 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37368 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43451 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21787 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65352 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20035 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78073 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16469 "Built successfully") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19191 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 10 flakes 2 failures; Uploaded test results; 13 flakes 1 failures; Compiled WebKit (warnings); layout-tests; Found 1 new test failure: http/wpt/mediarecorder/pause-recording.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65398 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16516 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62224 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64661 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15978 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12884 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6529 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47447 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2231 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48516 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49804 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48259 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/fido/CtapHidDriver.cpp:
```diff
@@ -62,8 +62,9 @@ void CtapHidDriverWorker::transact(fido::FidoHidMessage&& requestMessage, Messag
+    connection->invalidateCache();
+    connection->send(m_requestMessage->popNextPacket(), [weakThis = WeakPtr { *this }](HidConnection::DataSent sent) mutable {
         ASSERT(RunLoop::isMain());
         if (!weakThis)
```
Isn't `this` RefCounted? If so, I'd expect us to convert the weakThis into a `protectedThis` RefPtr here.

basuke replied:
Not yet. That's the task of next patch.

cdumez replied:
Yes, I'm actually working on it now :D

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/fido/CtapHidDriver.cpp:
```diff
@@ -83,7 +84,7 @@ void CtapHidDriverWorker::write(HidConnection::DataSent sent)
-        m_connection->registerDataReceivedCallback([weakThis = WeakPtr { *this }](Vector<uint8_t>&& data) mutable {
+        protectedConnection()->registerDataReceivedCallback([weakThis = WeakPtr { *this }](Vector<uint8_t>&& data) mutable {
             ASSERT(RunLoop::isMain());
             if (!weakThis)
```
ditto

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/fido/CtapHidDriver.cpp:
```diff
@@ -92,7 +93,7 @@ void CtapHidDriverWorker::write(HidConnection::DataSent sent)
-    m_connection->send(m_requestMessage->popNextPacket(), [weakThis = WeakPtr { *this }](HidConnection::DataSent sent) mutable {
+    protectedConnection()->send(m_requestMessage->popNextPacket(), [weakThis = WeakPtr { *this }](HidConnection::DataSent sent) mutable {
         ASSERT(RunLoop::isMain());
         if (!weakThis)
```
ditto

basuke committed:
[IsDeprecatedWeakRefSmartPointerException] Make more WebAuthentication classes RefCounted
https://bugs.webkit.org/show_bug.cgi?id=281637

Reviewed by Chris Dumez.

HidConnection class and its subclasses in WebAuthentication are managed as WeakPtr so they should be ref counted.

* Source/WebKit/UIProcess/WebAuthentication/Cocoa/HidConnection.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/HidConnection.mm:
(WebKit::HidConnection::create):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/HidService.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/HidService.mm:
(WebKit::HidService::createHidConnection const):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidConnection.cpp:
(WebKit::MockHidConnection::create):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidConnection.h:
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidService.cpp:
(WebKit::MockHidService::createHidConnection const):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidService.h:
* Source/WebKit/UIProcess/WebAuthentication/Virtual/VirtualHidConnection.cpp:
(WebKit::VirtualHidConnection::create):
* Source/WebKit/UIProcess/WebAuthentication/Virtual/VirtualHidConnection.h:
* Source/WebKit/UIProcess/WebAuthentication/Virtual/VirtualService.mm:
(WebKit::VirtualService::startDiscoveryInternal):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapHidDriver.cpp:
(WebKit::CtapHidDriverWorker::CtapHidDriverWorker):
(WebKit::CtapHidDriverWorker::~CtapHidDriverWorker):
(WebKit::CtapHidDriverWorker::transact):
(WebKit::CtapHidDriverWorker::write):
(WebKit::CtapHidDriverWorker::reset):
(WebKit::CtapHidDriverWorker::cancel):
(WebKit::CtapHidDriver::CtapHidDriver):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapHidDriver.h:
(WebKit::CtapHidDriverWorker::protectedConnection):

Canonical link: https://commits.webkit.org/285345@main
